### ISSUE TITLE
history: occasional white-balance inactivation when switching images

### ISSUE THREAD

**Issue by ralfbrown** [darktable-org/darktable#16828]
### Describe the bug

While blasting through 5000 images at around 180/hour last week, I noticed for about every 30-40 images on advancing to the next image (space bar) that the image I was leaving would briefly turn blue, and in some cases the red "trouble" triangle would momentarily show on the white balance and color calibration modules.  In addition, there's a trouble message in the console:
```
  2074.6151 Trouble: [white balance] white balance disabled but required (DSF89551.nef 376079)
  2074.6153 Trouble: [color calibration] white balance missing for color calibration (DSF89551.nef 376079)
```
It was either the main center image or the nav window plus filmstrip thumbnail which turned blue for a brief instant; I don't recall both doing so for the same image except when the effect was permanent (see below).  When the center image turned blue, it was replaced by the image I was switching to after a second or so.  When it was nav+thumbnail, sometimes the thumbnail blipped blue before the filmstrip advanced (so the outgoing image was still in center position and highlighted) and sometimes after (so the thumbnail to the left of the "active image" glitched).

Looks like a race condition on the image history, but it's mostly self-correcting - in all but two of the estimated 150 occurrences, the image and its history were fine after the brief visual glitch.  On those other two where the image stayed blue, the white balance module showed the default of "as shot to reference", but wasn't actually doing that operation.  In one case, switching to "reference" and then back resolved the trouble messages and blue display, in the other it was resetting the white balance module.

### Steps to reproduce

Not sure there's a specific reproducer other than going through edits as fast as possible, using lots of presets, styles, and sliders bound to keystroke shortcuts, as well as the "compress history" operation bound to a keystroke.  I mostly did a compress just before advancing to the next image, or next-to-last operation followed by hitting a digit to adjust the star rating.

### Expected behavior

no glitches on the history or displayed images

### Where did you obtain darktable from?

self compiled

### darktable version

recent 4.7.0 master versions since early May

### What OS are you using?

Linux

### What is the version of your OS?

Mageia 9

### Describe your system?

AMD Threadripper 3970X (32C64T), 64 GB RAM

### Are you using OpenCL GPU in darktable?

No

**Comment by ralfbrown** [comment]
Yeah, hard to reproduce is the usual problem with race conditions.

About a decade ago, I had an occasional glitch with a concurrent hash map I was implementing.  I wound up being able to make it happen on about 1/3 of test runs by cranking up the number of parallel threads hitting the map as fast as they could to 1536 (on a machine with 12 hardware threads).  That was finally enough to let me track it down.  Unfortunately we can't do that here since there's basically only a single writer and reader involved (updating the active image and refreshing the image processing/display).

**Comment by jenshannoschwalm** [comment]
Haven't seen this yet. For me the most likely race condition would be on `dev->image_storage` or `dev->chroma`

**Comment by jenshannoschwalm** [comment]
Had a look through the code and there is one thing suspicious at `views/darkroom.c L. 777`.

That seems to be wrong there, should be after L. 852 ? Would you check?

**Comment by ralfbrown** [comment]
Attempted fix was not successful.
